rs/orchestrator/src/upgrade.rs

```diff
@@ -136,20 +140,69 @@ impl Upgrade {
         // Determine the subnet_id using the local CUP.
-        let (subnet_id, local_cup_proto, local_cup) =
-            if let Some(proto) = self.cup_provider.get_local_cup_proto() {
-                let cup = CatchUpPackage::try_from(&proto).expect("deserializing CUP failed");
-                let subnet_id =
-                    get_subnet_id(&*self.registry.registry_client, &cup).map_err(|err| {
-                        OrchestratorError::UpgradeError(format!(
-                            "Couldn't determine the subnet id: {:?}",
-                            err
-                        ))
-                    })?;
-                (subnet_id, Some(proto), Some(cup))
-            } else {
-                // No local CUP found, check registry
-                match self.registry.get_subnet_id(latest_registry_version) {
+        let (subnet_id, local_cup_proto, local_cup) = {
+            let maybe_proto = self.cup_provider.get_local_cup_proto();
+            let maybe_cup = maybe_proto.as_ref().and_then(|proto| {
+                CatchUpPackage::try_from(proto)
+                    .map_err(|err| {
+                        error!(self.logger, "Failed to deserialize CatchUpPackage: {}", err);
+                    })
+                    .ok()
+            });
+
+            match (&maybe_cup, &maybe_proto) {
+                (Some(cup), _) => {
+                    let subnet_id =
+                        get_subnet_id(&*self.registry.registry_client, cup).map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+                    (subnet_id, maybe_proto, maybe_cup)
+                }
+                (None, Some(proto)) => {
+                    // We found a local CUP proto that we can't deserialize. This may only happen
+                    // if this is the first CUP we are reading on a new replica version after an
+                    // upgrade. This means we have to be an assigned node, otherwise we would have
+                    // left the subnet and deleted the CUP before upgrading to this version.
+                    // The only way to leave this branch is via subnet recovery.
+
+                    // Try to find the subnet ID by deserializing only the NiDkgId. If it fails
+                    // we will have to recover using failover nodes.
+                    let nidkg_id: NiDkgId = try_from_option_field(proto.signer.clone(), "NiDkgId")
+                        .map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't deserialize NiDkgId to determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+
+                    let subnet_id = match nidkg_id.target_subnet {
```

Comment:
What if in case where we cannot deserialize the CUP we, like in case of unassigned nodes, simply pull the subnet id from the registry? What could possible go wrong?

Reply:
Thanks for the explanation Leo!

Reply:
I thought a bit more about this. I guess my question was more within a specific context: once we cannot deserialize the CUP and are in the "disaster" branch above, we definitely shouldn't interpret the registry values as _outside_ of the disaster branch. 

Concretely: imagine we end up in the disaster branch (fail to deserialize the CUP) and the we read the subnet membership from the registry:

- if the conclusion is we're unassigned => do nothing, wait for manual upgrade,
- if we get a subnet id, check if it's latest registry CUP is at a _lower_ registry version than the upgrade that broke us.

Plase note I absolutely do not think this is better that relying on being able to deserialize the NiDKG data, I'm just exploring the entire solution space.

Reply:
PS: I guess we need to check two things: latest registry version touching the `SubnetRecord` and the corresponding CUP contents field.

Reply:
> if the conclusion is we're unassigned => do nothing, wait for manual upgrade,

That makes sense, I guess one downside would be that we would still have to re-provision these nodes that were unassigned.

> if we get a subnet id

Note that this may not be the subnet Id of the subnet that we got the broken CUP from, as the node may have been re-assigned.

> check if it's latest registry CUP is at a lower registry version than the upgrade that broke us

What signal would that give use exactly? If the upgrade to the incompatible version was triggered by a registry CUP, the registry version of that registry CUP would be higher than the version of the upgrade. If it was triggered by a normal peer CUP, then the registry version of the CUP is likely lower than the latest upgrade.

If the subnet ID returned by the registry is different from the subnet that we got the broken CUP on, then we don't know the registry version of the upgrade that broke us, because we no longer know which subnet we were on when the upgrade happened.